### ISSUE TITLE
How to group `db.run` calls into a transaction.

### ISSUE THREAD

**Issue by ghost** [TryGhost/node-sqlite3#907]
The docs mention that only the top level `db.run` calls are serialised:

https://github.com/mapbox/node-sqlite3/wiki/Control-Flow#databaseserializecallback
![image](https://user-images.githubusercontent.com/23221295/32794821-d72c4084-c961-11e7-91bd-1fcadf39f2ec.png)

Seeing as:

- I need to manage transactions manually with `BEGIN` and `END` calls.
- Each call is asynchronous; nested `db.run` callbacks in a single `db.serialize` are not serialized.

How can I prevent other writes to the database interfering with my transaction?

In other words, how can I ensure that when my code hits a `BEGIN` statement I can group writes together and prevent other writes not in that group interfering?

There is `db.exec`:
https://github.com/mapbox/node-sqlite3/wiki/API#databaseexecsql-callback

But how would I get the last insert ID reliably and use it for subsequent writes inside the same transaction?

Thanks.

Related Issues:

- https://github.com/mapbox/node-sqlite3/issues/3#issuecomment-7038796

**Comment by wmertens** [comment]
Sqlite3-the-library only runs a single query at a time per connection. There won't be an issue with racing, sqlite3-the-node-module doesn't use a connection pool.

You can run all those statements without the `db.serialize` and it will still work.

Anyway, you can nest `db.serialize` if you don't like that approach :)

For BEGIN/END, it would be best to run the transaction in a separate connection.

Also, this is more of a Stack Overflow question - I propose you close this.

**Comment by enzomd** [comment]
> You can run all those statements without the db.serialize and it will still work.

Does this mean the highlighted comment from the example code is wrong? (Two db.run queries are always run in order).

**Comment by wmertens** [comment]
> Does this mean the highlighted comment from the example code is wrong? (Two db.run queries are always run in order).

IMHO yes. See also #408 

> So you propose to _create a new connection to the db file for every transaction_, and only share the connection var/ref with the functions that write to that specific transaction (instead of having one global connection var/ref, and queueing writes to transactions so that unrelated queries are not run whilst a transaction is open)?

Correct.

> I think transaction semantics should be clear and have an API - previous issues for node-sqlite3 also mention this.

The problem with transactions and async single-threaded JS is that it is impossible to know which calls "belong" to the transaction. The only way to be sure is to use a different connection and pass it around to all functions cooperating in the transaction.

So I don't think this module can help with those semantics.

Perhaps some integration with https://github.com/audreyt/node-webworker-threads would work; node-sqlite3 could manage a pool of workers that each have their own connection; multi-threaded access to sqlite works fine but will mean copying data between worker threads and the main thread.

(Note, I'm not the maintainer of this project)

**Comment by troppoli** [comment]
I think that there are two reasons people will reach for Serialize. The first is that they want the statements run in order, the second is that they want them run in isolation (though it apparently won't help with that!). 

> The problem with transactions and async single-threaded JS is that it is impossible to know which calls "belong" to the transaction. The only way to be sure is to use a different connection and pass it around to all functions cooperating in the transaction.


It seems like there's a certain level of context that the module is aware of because the calls from within 'serialize' are known to be together...  

What I'm looking for is for another variant of serialize that queues all statements that are not from the callback until this callback is concluded. From there, I can create a meaningful transaction.

**Comment by daniellockyer** [comment]
Closing in favor of https://github.com/TryGhost/node-sqlite3/issues/304 re: transactions
